### PR TITLE
Merging tock process console with tockloader rs

### PR DESCRIPTION
**Pull request by NegrilaRares**

### CODE REVIEW

File: Cargo.toml
Comment (george-cosma) [inline Cargo.toml:1]: I was initially going to suggest into making this a [Root Package (workspace)](https://doc.rust-lang.org/cargo/reference/workspaces.html#root-package) rather than a [Virtual Workspace](https://doc.rust-lang.org/cargo/reference/workspaces.html#virtual-workspace). However, if we want to split our code into both a library and CLI, then it might be worth follow [probe-rs's example](https://github.com/probe-rs/probe-rs/blob/master/Cargo.toml) which also uses this type of Cargo.toml. It has `probe-rs` as the library and `probe-rs-tools` as the CLI tool.

What do you think?

Note: for the purposes of this PR I don't want to split it YET. Let's do that in another PR, after this one gets merged.

File: tock-process-console/loguri.log
Comment (george-cosma) [inline tock-process-console/loguri.log:1]: This needs to be added to .gitignore

File: tock-process-console/board-mockk.log
Comment (george-cosma) [inline tock-process-console/board-mockk.log:1]: This needs to be added to .gitignore

File: tock-process-console/src/.DS_Store
Comment (george-cosma) [inline tock-process-console/src/.DS_Store:1]: This needs to be added to .gitignore

File: tock-process-console/src/board/connection.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    decoder: Decoder,
+    decoded_receiver: UnboundedReceiver<BoardMessage>,
+}
+
+impl ConnectionHandler {
+    pub async fn new(
+        tty: &str,
+    ) -> Result<(UnboundedReceiver<Event>, UnboundedSender<Bytes>), Error> {
+        let mut port: SerialStream = tokio_serial::new(tty, 115200).open_native_async()?;
+        port.set_exclusive(false)?;
```
Comment (george-cosma) [inline tock-process-console/src/board/connection.rs:28]: This doesn't compile on Windows. add a `#cfg` directive to be excluded from windows

File: tock-process-console/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+mod board;
+mod state_store;
+mod termination;
+mod ui_management;
+
+use state_store::StateStore;
+use termination::{create_terminator, Interrupted};
+use ui_management::UiManager;
+
+pub async extern fn run() ->  anyhow::Result<()> {
```
Comment (george-cosma) [inline tock-process-console/src/lib.rs:10]: Why `extern`?

File: tock-process-console/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        println!("Exited due to an unexpected error");
+    }
+
+    Ok(())
+}
+
+
+
+#[cfg(test)]
+mod tests {
```
Comment (george-cosma) [inline tock-process-console/src/lib.rs:35]: Is this intentional? If not, I think this should be deleted

File: tock-process-console/src/ui_management/.DS_Store
Comment (george-cosma) [inline tock-process-console/src/ui_management/.DS_Store:1]: This needs to be added to .gitignore

File: tockloader/LICENSE
Comment (george-cosma) [inline tockloader/LICENSE:1]: LICENSE should remain in the root of the repository

File: tockloader/Makefile
Comment (george-cosma) [inline tockloader/Makefile:1]: This should probably remain in the root of the repository, but we will need to make another PR to set up github actions

File: tockloader/src/errors.rs [hunk 4]
```diff
@@ -40,7 +40,7 @@ impl fmt::Display for CLIError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
             CLIError::MultipleInterfaces => {
-                f.write_str("At most one of the following tranport interfaces may be used: '--serial', '--openocd', '-jlink'")
+                f.write_str("At most one of the following tranport interfaces may be used: '--serial'")
```
Comment (george-cosma) [inline tockloader/src/errors.rs:43]: This error is likely no longer relevant. However, we might need it in the future. Add a comment 
```rust
// TODO (github-username-here): Do we need still need this error?
f.write_str(...)
```

File: tockloader/src/interfaces/serial.rs [hunk 5]
```diff
@@ -58,4 +64,4 @@ fn clean_port_path(port: String) -> String {
     } else {
         port
     }
-}
+}
```
Comment (george-cosma) [inline tockloader/src/interfaces/serial.rs:67]: Run `cargo fmt`

File: tockloader/src/main.rs [hunk 6]
```diff
@@ -25,8 +31,15 @@ async fn run() -> Result<(), TockloaderError> {
     match matches.subcommand() {
         Some(("listen", sub_matches)) => {
             let mut interface = build_interface(sub_matches)?;
```
Comment (george-cosma) [inline tockloader/src/main.rs:33]: I don't think we need to create an interface for this. Especially since the `tock-process-console` already deals with serial communications. We will likely also need to merge these somehow, but I would leave them separate for now

File: tockloader/tools/run_clippy.sh
Comment (george-cosma) [inline tockloader/tools/run_clippy.sh:1]: This (`tools`) should remain in root

File: tockloader/tools/run_fmt_check.sh
Comment (george-cosma) [inline tockloader/tools/run_fmt_check.sh:1]: This (`tools`) should remain in root

**Review by george-cosma — CHANGES_REQUESTED** [review]

File: tockloader/src/main.rs (hunk 6 above)
Comment (NegrilaRares) [reply tockloader/src/main.rs:33]: was previously using build_interface as a pre-ratatui way of viewing all your ports to know what to chose for the input for ratatui since ill be needing this ill prob get rid of it when i complete the ratatui serial changes.

but it is indeed not needed just temporarily convenient for me.

File: tock-process-console/src/board/connection.rs (hunk 1 above)
Comment (NegrilaRares) [reply tock-process-console/src/board/connection.rs:28]: i opted for a cfg! since its just one line

**Comment by NegrilaRares** [comment]
lots of things will need to be tweaked and deleted later on

File: tock-process-console/.gitignore [hunk 7]
```diff
@@ -0,0 +1,6 @@
+/target
+/board-mock.log
+/board-mockk.log
+/loguri.log
+/src/.DS_Store
+/src/ui_management/.DS_Store
```
Comment (george-cosma) [inline tock-process-console/.gitignore:6]: instead of adding each individual file, add a glob:
```rust
*.log
.DS_Store
```
Notice the lack `/` at the start. This means any files that ends in .log or any file that is ".DS_Store" will be automatically ignored

File: tock-process-console/board-mock.log
Comment (george-cosma) [inline tock-process-console/board-mock.log:1]: You have removed the files with gitignore, but they are still in the PR.

**Review by george-cosma — CHANGES_REQUESTED** [review]

File: tockloader/tools/run_fmt_check.sh
Comment (george-cosma) [reply tockloader/tools/run_fmt_check.sh:1]: This is not yet solved

**Comment by george-cosma** [comment]
LGTM. I'll wait a bit, actually since I want to figure out this License stuff first.

**Comment by george-cosma** [comment]
👍

**Review by george-cosma — APPROVED** [review]
